### PR TITLE
Remove SmartFormat and roll our own code instead

### PR DESCRIPTION
**Pull request by liadtal**

### CODE REVIEW

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs [hunk 1]
```diff
@@ -35,5 +38,85 @@ public static string Hash(this string s)
...
+        ///      => "User john@contoso.com with role Administrator created at Monday, June 15, 2009
+        ///    "Value decreased at {DecreasePercent:F2}% for metric {MetricName}".FormatWithObjectProperties(metric)
+        ///      => "Value decreased at 22.10% for metric TotalRevenue"
+        ///    "The array {{Length}} is {Length}".FormatStringWithObjectProperties(new int[] {1,2,3,4})
+        ///      => "The array {Length} is 4"
+        /// </summary>
+        /// <param name="format">The format string</param>
+        /// <param name="source">The source object</param>
+        /// <returns>The result string</returns>
+        public static string FormatWithObjectProperties(this string format, object source)
```
Comment (yanivid3) [inline src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: >FormatWithObjectProperties [](start = 29, length = 26)

is it possible to perform injections?
Not sure I have a good example...An evil user that provides info which will appear in UI.

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs [hunk 2]
```diff
@@ -35,5 +38,85 @@ public static string Hash(this string s)
...
+                        newFormat.Append("{{");
+                        i++;
+                    }
+                    else
+                    {
+                        // Find the closing brace
+                        int closingBraceIndex = format.IndexOf('}', i);
+                        if (closingBraceIndex < 0)
+                        {
+                            throw new FormatException($"Encountered an opening brace without a closing brace in format string: {format}");
```
Comment (EranGon) [inline src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: >{format} [](start = 127, length = 8)

'{format}'

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs [hunk 3]
```diff
@@ -35,5 +38,85 @@ public static string Hash(this string s)
...
+            // Parse the format string
+            StringBuilder newFormat = new StringBuilder();
+            List<object> values = new List<object>();
+            for (int i = 0; i < format.Length; i++)
+            {
+                if (format[i] == '{')
+                {
+                    if (i < format.Length - 1 && format[i + 1] == '{')
+                    {
+                        // Escaped brace - add it as-is, and skip
```
Comment (yanivid3) [inline src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: >and skip [](start = 57, length = 8)

test this case :)

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs (hunk 2 above)
Comment (yanivid3) [inline src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: >brace without [](start = 79, length = 13)

test this case :)

Comment (EranGon) [inline src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: >string [](start = 119, length = 6)

also add the index of the openning {

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs [hunk 4]
```diff
@@ -35,5 +38,85 @@ public static string Hash(this string s)
...
+                        newFormat.Append("{{");
+                        i++;
+                    }
+                    else
+                    {
+                        // Find the closing brace
+                        int closingBraceIndex = format.IndexOf('}', i);
```
Comment (EranGon) [inline src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: >int closingBraceIndex = format.IndexOf('}', i); [](start = 24, length = 47)

what about escaped }?

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs [hunk 5]
```diff
@@ -35,5 +38,85 @@ public static string Hash(this string s)
...
+                        newFormat.Append("{" + values.Count + specifier + "}");
+
+                        // Get the property from the object and add it to the values list
+                        PropertyInfo propertyInfo = source.GetType().GetProperty(name);
+                        if (propertyInfo == null)
+                        {
+                            throw new FormatException($"Property {name} not found in object of type {source.GetType().Name}");
+                        }
+
+                        values.Add(propertyInfo.GetValue(source));
```
Comment (EranGon) [inline src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: please look at the code sample i attached to the task - you're much better off using expressions for stuff like:
"Disk space about to run out in {MyTime.TotalMinutes} minutes"
or:
"Found {NumOfErrors > 0 ? NumOfErrors.ToString() : "no"} errors"

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs (hunk 3 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: It is tested in the unit tests

---
In reply to: [185494523](https://github.com/liadtal/Smart-Alerts/pull/145#discussion_r185494523) [](ancestors = 185494523)

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs (hunk 5 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: I saw that code sample - several reasons I didn't use it:
1. It doesn't handle escaped strings, and other corner cases.
2. It uses System.Linq.Dynamic, which is not a standard DLL. The Nuget that contains it ([https://github.com/kahanu/System.Linq.Dynamic](https://github.com/kahanu/System.Linq.Dynamic)) was uploaded by someone who claims not to take ownership to it. I didn't want to rely on it.
3. We can use, instead, Roslyn to compile the string in brackets. But that can introduce other problems and dependencies.

---
In reply to: [185497851](https://github.com/liadtal/Smart-Alerts/pull/145#discussion_r185497851) [](ancestors = 185497851)

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs (hunk 1 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: I don't think adding this increases the vulnerability that we already have - we are running a custom detector. Can you think of an example?

---
In reply to: [185494062](https://github.com/liadtal/Smart-Alerts/pull/145#discussion_r185494062) [](ancestors = 185494062)

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs (hunk 4 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: If we have a }} after an opening { then this is a format exception, and string.Format (below) will throw an exception in this scenario.

---
In reply to: [185495741](https://github.com/liadtal/Smart-Alerts/pull/145#discussion_r185495741) [](ancestors = 185495741)

Comment (EranGon) [reply src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: not sure this should be the case, but I guess this is irrelevant when we move to expressions

---
In reply to: [185576752](https://github.com/liadtal/Smart-Alerts/pull/145#discussion_r185576752) [](ancestors = 185576752,185495741)

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs [hunk 6]
```diff
@@ -35,5 +38,23 @@ public static string Hash(this string s)
...
+        ///    "Found {NumAffectedMachines} affected {(NumAffectedMachines == 1 ? \"machine\" : \"machines\")}"
+        ///      => "Found 1 affected machine", "Found 3 affected machines"
+        /// </summary>
+        /// <param name="interpolatedStringDefinition">The interpolated string definition.</param>
+        /// <param name="source">The object, whose properties can appear in the interpolated string definition. The object type definition must be public.</param>
+        /// <returns>The evaluated result string</returns>
+        public static string EvaluateInterpolatedString(string interpolatedStringDefinition, object source)
+        {
+            string code = "$\"" + interpolatedStringDefinition + "\"";
+            return CSharpScript.EvaluateAsync<string>(code, globals: source, globalsType: source.GetType()).Result;
```
Comment (EranGon) [inline src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: > return CSharpScript.EvaluateAsync<string>(code, globals: source, globalsType: source.GetType()).Result; [](start = 11, length = 104)

hot damn!

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs (hunk 1 above)
Comment (yanivid3) [reply src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: No :) 
BUT i'm not so  evil...

---
In reply to: [185534075](https://github.com/liadtal/Smart-Alerts/pull/145#discussion_r185534075) [](ancestors = 185534075,185494062)

**Pull request merged**